src/main/java/org/reaktivity/nukleus/mqtt/internal/stream/MqttServerFactory.java

```diff
@@ -837,10 +931,93 @@ private int decodeSession(
         }
     }
 
+    private final class Subscription
+    {
+        private final MqttServer server;
+        private final int fullReasonCodesMask;
+        private final List<Byte> reasonCodes;
+
+        private int reasonCodesMask;
+
+        private Subscription(
+            MqttServer server,
+            int reasonCodesCount)
+        {
+            this.server = server;
+            this.reasonCodes = new ArrayList<>();
+            for (int i = 0; i < reasonCodesCount; i++)
+            {
+                reasonCodes.add((byte) 0x8f);
+            }
+            this.fullReasonCodesMask = 1 << (reasonCodesCount - 1);
+        }
+
+        private void onSubscribeFailed(
+            int packetId,
+            int ackIndex)
+        {
+            final int bit = 1 << ackIndex;
+            reasonCodes.set(ackIndex, (byte) 0x8f);
+            reasonCodesMask |= bit;
+            if ((reasonCodesMask & fullReasonCodesMask) == fullReasonCodesMask)
+            {
+                doMqttSuback(packetId);
+            }
+        }
+
+        private void onSubscribeSucceeded(
+            int packetId,
+            int ackIndex)
+        {
+            final int bit = 1 << ackIndex;
+            reasonCodes.set(ackIndex, (byte) 0x00);
+            reasonCodesMask |= bit;
+            if ((reasonCodesMask & fullReasonCodesMask) == fullReasonCodesMask)
+            {
+                doMqttSuback(packetId);
+            }
+        }
```

Comment:
Both `onSubscribeFailed` and `onSubscribeSucceeded` should defer to `onSubscribeCompleted(..., ackIndex, reasonCode)`, passing `reasonCode` `0x00` for succeeded and `0x8f` for failed.